src/components/Dashboard/Dashboard.tsx

```diff
@@ -11,81 +14,213 @@ import {
  * @public
  */
 export interface IDashboard {
+  /**
+   * The widgets to make available in this Dashboard.
+   */
   widgets: IWidget[];
+  /**
+   * Any initial preferences that should be set for this user, in case you save users’ preferences remotely.
+   */
+  preferences?: IDashboardPreferences;
+  /**
+   * Set this `cacheKey` to save user’s preferences on their local clients.
+   */
+  cacheKey?: TCacheKey;
+  /**
+   * A Dashboard will emit onInteraction payloads when the user updates any preferences.
+   */
+  onInteraction?: (interaction: TDashboardInteraction) => void;
 }
 
-const toolbarConfig = {
-  actionGroups: {
-    h1: {
-      edit: { title: "Edit dashboard", icon: "Edit" },
-    },
-  },
-  filters: [],
-  find: false,
-};
+/**
+ * A user’s preferences for the particular Dashboard component. For a given Dashboard instance,
+ * a user may set certain widgets to be shown or hidden.
+ * @public
+ */
+export interface IDashboardPreferences {
+  // widgetOrder: string[];
+  widgetSettings: {
+    [widgetKey: string]: {
+      display: boolean;
+    };
+  };
+}
 
 /**
+ * A Dashboard will emit onInteraction payloads when the user updates any preferences.
  * @public
  */
-export function Dashboard({ widgets }: IDashboard) {
+export type TDashboardInteraction = IDashboardInteractionUpdatePreferences;
+
+/**
+ * The preferences update payload carries the preferences the developer should store for the user,
+ * if appropriate.
+ * @public
+ */
+export interface IDashboardInteractionUpdatePreferences {
+  event: "update";
+  target: "preferences";
+  preferences: IDashboardPreferences;
+}
+
+const emptyPrefs = { widgetSettings: {} };
+
+/*
+ * This method returns the same stored preferences for any Dashboard.
+ */
+const getStoredPrefs = (cacheKey: TCacheKey) =>
+  cacheKey
+    ? (() => {
+        const storedPrefs = window.localStorage.getItem(cacheKey);
+        return storedPrefs ? JSON.parse(storedPrefs) : false;
+      })() || emptyPrefs
+    : (emptyPrefs as IDashboardPreferences);
+
+/**
+ * @public
+ */
+export function Dashboard({
+  widgets,
+  preferences,
+  cacheKey,
+  onInteraction,
+}: IDashboard) {
+  const [sidebarOpen, setSidebarOpen] = useState<boolean>(false);
+  const closeSidebar = () => setSidebarOpen(false);
+
+  const localStorageKey = cacheKey
+    ? `@fluentui/react-teams__${cacheKey}`
+    : undefined;
+
+  const [preferencesState, setPreferencesState] = useState<
+    IDashboardPreferences
+  >(
+    (() => {
+      return cloneDeep(
+        widgets.reduce((loadedPrefs, { id }) => {
+          return set(
+            loadedPrefs,
+            `widgetSettings.${id}.display`,
+            get(loadedPrefs, `widgetSettings.${id}.display`, true)
+          );
+        }, preferences || getStoredPrefs(localStorageKey))
+      );
+    })()
+  );
+
+  const updatePreferences = (nextPreferences: IDashboardPreferences) => {
+    setPreferencesState(cloneDeep(nextPreferences));
```

Comment:
Not specific to this PR, but a general comment: you may want to look into incorporating something like https://github.com/immerjs/immer to avoid deep clones.

Reply:
Nice! I like this idea for sure. There are other places in this library that could use that optimization.